src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DeconstructionAssignmentOperator.cs

```diff
@@ -320,6 +350,43 @@ internal static DeconstructionSideEffects GetInstance()
                 return result;
             }
 
+            internal BoundExpression PopLast()
+            {
+                if (assignments.Count != 0)
+                {
+                    var last = assignments.Last();
+                    assignments.RemoveLast();
+                    return last;
+                }
+                if (conversions.Count != 0)
+                {
+                    var last = conversions.Last();
+                    conversions.RemoveLast();
+                    return last;
+                }
+                if (deconstructions.Count != 0)
+                {
+                    var last = deconstructions.Last();
+                    deconstructions.RemoveLast();
+                    return last;
+                }
+                if (init.Count != 0)
+                {
+                    var last = init.Last();
+                    init.RemoveLast();
+                    return last;
+                }
+                return null;
+            }
```

Comment:
There are five separate cases in `PopLast`. Are we testing all of these? Alternatively, would this be simpler if `ToImmutableAndFree()` is split into separate `GetExpressions` and `Free` methods, and the caller handles getting the last expression?
```
void GetExpressions(ArrayBuilder<Expression> builder)
{
    builder.AddRange(init);
    builder.AddRange(deconstructions);
    builder.AddRange(conversions);
    builder.AddRange(assignments);
}
```
In the caller:
```
var builder = ArrayBuilder<Expression>.GetInstance();
effects.GetExpressions(builder);
effects.Free();
if (isUsed)
{
    result = _factory._factory.Sequence(temps.ToImmutableAndFree(), builder.ToImmutable(), returnValue);
}
else if (builder.Count == 0)
{
    result = null;
}
else
{
    var last = builder.Last();
    builder.RemoveLast();
    result = _factory._factory.Sequence(temps.ToImmutableAndFree(), builder.ToImmutable(), last);
}
builder.Free();
```